ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -555,7 +554,11 @@ func (r *RayClusterReconciler) reconcilePods(instance *rayv1alpha1.RayCluster) e
 					}
 					r.Log.Info("reconcilePods", "unable to delete worker ", pod.Name)
 				} else {
-					diff++
+					// For example, the failed Pod (Status: corev1.PodFailed) is not counted in the `runningPods` variable.
+					// Therefore, we should not update `diff` when we delete a failed Pod.
+					if isPodRunningOrPendingAndNotDeleting(pod) {
```

Comment:
I cannot understand the point you are trying to make. We still need to calculate the number of Pods in `WorkersToDelete` that overlap with `runningPods`. Would you mind providing more details about the point you're trying to make? We can continue the discussion here or chat offline.

Reply:
Sync with @Yicheng-Lu-llll offline: 

[Yicheng's point]

This loop will send `Delete` requests to the Kubernetes API Server to delete Pods in `WorkersToDelete'` Afterward, we can list the worker Pods so that we do not need to maintain the `diff` within the loop.

https://github.com/ray-project/kuberay/blob/c4201350ef13598490031f936b945565dcde9dd5/ray-operator/controllers/ray/raycluster_controller.go#L472-L476

[Conclusion]

We cannot guarantee that the informer cache will be consistent with the Kubernetes API Server when listing worker Pods. Therefore, it is still necessary to calculate the `diff` within the loop.